qutip/qip/circuit.py

```diff
@@ -593,6 +593,13 @@ def add_circuit(self, qc, start=0):
         if self.N - start < qc.N:
             raise NotImplementedError("Targets exceed number of qubits.")
 
+        # Inherit the user gates
+        for user_gate in qc.user_gates:
+            if user_gate in self.user_gates and overwrite_user_gates:
+                self.user_gates[user_gate] = qc.user_gates[user_gate]
+            else:
+                self.user_gates[user_gate] = qc.user_gates[user_gate]
```

Comment:
They seem to do the same thing regardless of the if branch. Maybe you meant to do nothing if `overwrite_user_gates==False`?

Reply:
Ah, sorry, I've fixed the problem.

---

I just added a test into the `test_add_circuit`. I just found that the offset was ignored when adding user defined gates. Indeed, classifying the type of gates is unnecessary here. The arguments of the operators are well defined in the original circuit. I've cleaned up the messy branches.

By the way, is there a particular reason to set the default values of the targets and controls to `None` instead of empty lists? Since the unnecessary arguments are ignored in simulation, it seems loses nothing by setting the default values to empty sets. Since we are iterating over the targets and controls when we move the gates around, it should be simpler if we can iterate over the default value. But I'm not sure how big the change would be. The nullity checks exist in many places, and they have to be modified if we change the default values.

Reply:
> By the way, is there a particular reason to set the default values of the targets and controls to None instead of empty lists? 

I don't think there is a particular reason. I kind of agree that replacing them with an empty list is a better option. But it is gonna be a "large" scale change as it exists in many places (which also means that `None` was indeed a bad choice). So it is better done in the new qutip-qip rather than here.